amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -114,12 +126,107 @@ public void restartFrom(RecordsRetrieved recordsRetrieved) {
+                .equals(recordsDeliveryAck.batchUniqueIdentifier())) {
+            // It is now safe to remove the element
+            recordsDeliveryQueue.poll();
+            // Take action based on the time spent by the event in queue.
+            takeDelayedDeliveryActionIfRequired(shardId, recordsRetrievedContext.getEnqueueTimestamp(), log);
+            // Update current sequence number for the successfully delivered event.
+            currentSequenceNumber = ((FanoutRecordsRetrieved)recordsRetrievedContext.getRecordsRetrieved()).continuationSequenceNumber();
+            // Update the triggering flow for post scheduling upstream request.
+            flowToBeReturned = recordsRetrievedContext.getRecordFlow();
+            // Try scheduling the next event in the queue, if available.
+            if (recordsDeliveryQueue.peek() != null) {
+                subscriber.onNext(recordsDeliveryQueue.peek().getRecordsRetrieved());
+            }
+        } else {
+            // Check if the mismatched event belongs to active flow. If publisher receives an ack for a
+            // missing event in active flow, then it means the event was already acked or cleared
+            // from the queue due to a potential bug.
+            if (flow != null && recordsDeliveryAck.batchUniqueIdentifier().getFlowIdentifier()
+                    .equals(flow.getSubscribeToShardId())) {
+                log.error(
+                        "{}: Received unexpected ack for the active subscription {}. Throwing.  ",
+                        shardId, recordsDeliveryAck.batchUniqueIdentifier().getFlowIdentifier());
+                throw new IllegalStateException("Unexpected ack for the active subscription");
+            }
+            // Otherwise publisher received a stale ack.
+            else {
+                log.info("{}: Publisher received an ack for stale subscription {}. Ignoring.", shardId,
+                        recordsDeliveryAck.batchUniqueIdentifier().getFlowIdentifier());
+            }
+        }
+        return flowToBeReturned;
+    }
+
+    // This method is not thread-safe. You need to acquire a lock in the caller in order to execute this.
+    @VisibleForTesting
+    void bufferCurrentEventAndScheduleIfRequired(RecordsRetrieved recordsRetrieved, RecordFlow triggeringFlow) {
+        final RecordsRetrievedContext recordsRetrievedContext =
+                new RecordsRetrievedContext(recordsRetrieved, triggeringFlow, Instant.now());
+        try {
+            // Try enqueueing the RecordsRetrieved batch to the queue, which would throw exception on failure.
+            // Note: This does not block wait to enqueue.
+            recordsDeliveryQueue.add(recordsRetrievedContext);
+            // If the current batch is the only element in the queue, then try scheduling the event delivery.
+            if (recordsDeliveryQueue.size() == 1) {
```

Comment:
A question about this code path: So recordsReceived seems to be another code path that we receive the records without the ack? Are we still allowing the subscriber to schedule another record in this case?

Reply:
The SDK thread dispatches the records to our Publisher in a blocking manner. That is an event(T) will be dispatched only when the SDK thread sees the Future for event(T-1) is completed successfully.